arrows/klv/klv_demuxer.cxx

```diff
@@ -0,0 +1,431 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+/// \file
+/// \brief Implementation of KLV demuxer.
+
+#include "klv_0104_new.h"
+#include "klv_0601_new.h"
+#include "klv_1108.h"
+#include "klv_1108_metric_set.h"
+#include "klv_demuxer.h"
+
+#include <vital/logger/logger.h>
+#include <vital/range/iota.h>
+
+namespace kv = kwiver::vital;
+
+namespace kwiver {
+
+namespace arrows {
+
+namespace klv {
+
+namespace {
+
+// ----------------------------------------------------------------------------
+// Values are valid for 30 seconds if not explicitly overridden.
+constexpr uint64_t klv_packet_default_duration = 30000000;
+constexpr uint64_t klv_0104_default_duration = klv_packet_default_duration;
+constexpr uint64_t klv_0601_default_duration = klv_packet_default_duration;
+
+// ----------------------------------------------------------------------------
+uint64_t
+klv_find_or_insert_1108( klv_timeline& timeline,
+                         klv_local_set const& parent_set,
+                         klv_value const& metric_set_value )
+{
+  constexpr auto standard = KLV_PACKET_MISB_1108_LOCAL_SET;
+
+  // Invalid metric sets are all treated as unique and therefore an equivalent
+  // sub-timeline cannot be found
+  if( metric_set_value.valid() )
+  {
+    auto const& metric_set = metric_set_value.get< klv_local_set >();
+
+    // The METRIC_LOCAL_SET tag is used to get all the unique metrics being
+    // tracked.
+    auto const candidates =
+      timeline.find_all( standard, KLV_1108_METRIC_LOCAL_SET );
+    for( auto candidate : candidates )
+    {
+      // We are trying to find out if this candidate set of sub-timelines is
+      // tracking the same metric as the new one we want to insert
+      auto is_match = true;
+      auto const index = candidate.first.index;
+
+      // Either of these tags in the parent set being different would mean the
+      // new metric should be tracked separately
+      for( auto const tag : { KLV_1108_ASSESSMENT_POINT,
+                              KLV_1108_WINDOW_CORNERS_PACK } )
+      {
+        auto const it = timeline.find( standard, tag, index );
+
+        // All this to check whether the values are different, the complexity
+        // coming from the fact that we don't assume either value exists in the
+        // first place, since WINDOWS_CORNER_PACK is optional
+        if( it != timeline.end() && !it->second.empty() &&
+            ( it->second.begin()->value.empty() != !parent_set.has( tag ) ||
```

Comment:
This is another case of a double negative; i.e. `!= !foo`.  Is the logic
`( it->second.begin()->value.empty() == parent_set.has( tag )` not possible?  It seems more readable if it's possible.